Fix truncated messages when editing long messages.

kacf opened the pull request:
The problem manifests when you are trying to edit a long message. The
partial URL mode causes the request, and therefore the message, to be
truncated. A message with a 100 characters is more than enough to
trigger it.

Therefore switch to POST so we can include the long parameters in the
body instead. We take the opportunity to switch all appropriate
endpoints over to POST, which is a more correct way to use the API in
general.

It's a bit hairy to do POST requests since Purple does not support
them natively, but we just need to construct a small set of headers.

Signed-off-by: Kristian Amlie <kristian.amlie@northern.tech>

EionRobb commented:
Full url mode was causing issues for me when there was the json response error. Might depend on which server to connect to, or light have been resolved server side after the latest downtime 

I'd personally be a bit nervous of this change

kacf commented:
Is there a particular call I can make to test that? Is it just when I make an invalid request?

dylex commented:
The right solution to this is probably to switch to POST instead of GET calls (which would also be useful for a number of other calls as well), but since libpurple doesn't provide this (that I know of) we'd have to implement more of HTTP ourselves (or pull in libcurl or something).

EionRobb commented:
@dylex you can use the same `purple_util_fetch_url_request_len_with_account()` call to do posts, it's just a little more fiddly: https://github.com/EionRobb/purple-discord/blob/master/libdiscord.c#L1222-L1235

kacf commented:
Ok, thanks, I will try with POST.

kacf commented:
Thanks @EionRobb, the discord code was a big inspiration! I have now switched all appropriate calls over to the new `slack_api_post`, and for clarity I renamed the old `slack_api_call` to `slack_api_get`. Quite a bit bigger than the original one line fix, but definitely more correct!

kacf commented:
This does not merge cleanly with the `history` branch, so I also created #128.

dylex commented on slack-api.c:
```diff
@@ -97,33 +105,117 @@ static GString *slack_api_encode_url(SlackAccount *sa, const char *pfx, const ch
 	call->sa = sa;
 	call->callback = callback;
 	call->url = g_strdup(url);
+	if (request)
```
this if is unnecessary: `g_strdup(NULL) = NULL`

dylex reviewed (approved):
Thanks, this is great! Thanks for doing all this, I think it'll make things much better. I made a few minor comments, but I can take another closer pass and make these changes myself.

kacf commented:
Thanks for merging @dylex! Your suggestions make perfect sense, I was just unaware of them!
